rest/api.go

```diff
@@ -0,0 +1,243 @@
+	PositionLine *float64 `json:"positionLine,omitempty"`
+	// PositionChar for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+	PositionChar *float64 `json:"positionChar,omitempty"`
+}
+
+func (d *lspData) UnmarshalJSON(data []byte) error {
+	var tmp lspData
+	type plain lspData
+	if err := json.Unmarshal(data, (*plain)(&tmp)); err != nil {
+		return err
+	}
+	if len(tmp.Expr) == 0 {
+		return fmt.Errorf("promQL expression is not specified")
+	}
+	*d = tmp
+	return nil
+}
+
+func (d *lspData) returnPosition() (protocol.Position, error) {
+	if d.PositionLine == nil {
+		return protocol.Position{}, errors.New("positionLine is not specified")
+	}
+	if d.PositionChar == nil {
+		return protocol.Position{}, errors.New("positionChar is not specified")
+	}
+	return protocol.Position{
+		Line:      *d.PositionLine,
+		Character: *d.PositionChar,
+	}, nil
+}
+
+type API struct {
+	langServer    langserver.HeadlessServer
+	mdws          []middlewareFunc
+	enableMetrics bool
+}
+
+// NewLangServerAPI create a new instance of the Stateless API to use the LangServer through HTTP.
+//
+// If metadata is fetched from a remote Prometheus, the metadataService
+// implementation from the promql-langserver/prometheus package can be used,
+// otherwise you need to provide your own implementation of the interface.
+//
+// The provided Logger should be synchronized.
+//
+// In case "enableMetrics" is set to true, endpoint /metrics is then available and a middleware that instrument the different endpoints provided is instantiated.
+// Don't use it in case you have already in place such middleware.
+func NewLangServerAPI(ctx context.Context, metadataService promClient.MetadataService, logger log.Logger, enableMetrics bool) (*API, error) {
+	lgs, err := langserver.CreateHeadlessServer(ctx, metadataService, logger)
+	if err != nil {
+		return nil, err
+	}
+	mdws := []middlewareFunc{manageDocumentMiddleware(lgs)}
+	if enableMetrics {
+		apiMetric := newAPIMetrics()
+		prometheus.MustRegister(apiMetric)
+		mdws = append(mdws, apiMetric.instrumentHTTPRequest)
+	}
+	return &API{
+		langServer:    lgs,
+		mdws:          mdws,
+		enableMetrics: enableMetrics,
+	}, nil
+}
+
+// Register the API's endpoints in the given router.
+func (a *API) Register(r *route.Router, prefix string) {
+	r.Post(prefix+"/diagnostics", a.handle(a.diagnostics))
+	r.Post(prefix+"/completion", a.handle(a.completion))
+	r.Post(prefix+"/hover", a.handle(a.hover))
+	r.Post(prefix+"/signatureHelp", a.handle(a.signature))
+	if a.enableMetrics {
+		r.Get("/metrics", promhttp.Handler().ServeHTTP)
+	}
+}
+
+func (a *API) handle(h http.HandlerFunc) http.HandlerFunc {
+	endpoint := h
+	for _, mdw := range a.mdws {
+		endpoint = mdw(endpoint)
+	}
+	return endpoint
+}
+
+func (a *API) diagnostics(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	requestID, requestData, err := getRequestDataAndID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
```

Comment:
Use the `http.StatusXXX` constants here and elsewhere, instead of direct numeric codes (makes it easier for people to immediately understand the meaning of the error).